### PR TITLE
Batch up power level event when creating rooms

### PR DESCRIPTION
**Pull request by H-Shay**
Follow on work from #13800. Add the initial power level event to the group of events that are batch persisted when the room is created. Reviewable commit-by-commit.

### CODE REVIEW

File: synapse/event_auth.py [hunk 1]
```diff
@@ -15,7 +15,18 @@
```
Comment (MadLittleMods) [inline synapse/event_auth.py:15]: Some discussion in `#synapse-dev` about this PR: https://matrix.to/#/!vcyiEtMVHIhWXcJAfl:sw1v.org/$oaV7wGmCQOD-pByNgH6x1YhrVrqHBnEuKD76QQ7bClU?via=matrix.org&via=element.io&via=beeper.com

Comment (MadLittleMods) [reply synapse/event_auth.py:15]: Sorry, wrong PR

**Review by reivilibre — CHANGES_REQUESTED** [review]
This seems reasonable to me at first glance.

Though a point I wanted to check:
are these events persisted in a chain that all point to each other? I thought selecting the prev_events would happen in `create_event`, which seems to be happening before the events get persisted...
(Is there a test around that tests that the events created when creating a room are a chain?)

(edit: just thinking that it seems quite important for the power levels event to come before the others)

It also looks like you've collided with the push rules evaluator port to Rust, so a big chunk of your changes need porting :(

**Comment by erikjohnston** [comment]
I think it might be helpful to split the batching up of push rules into a separate PR too? A quick glance looks like there are a few different things going on in here.

**Comment by H-Shay** [comment]
> Though a point I wanted to check: are these events persisted in a chain that all point to each other? I thought selecting the prev_events would happen in `create_event`, which seems to be happening before the events get persisted... (Is there a test around that tests that the events created when creating a room are a chain?)
The events are indeed persisted in a chain pointing to one another. While there is not an explicit test verifying this, a number of existing tests implicitly rely on this behavior and break in arcane ways when this assumption is violated. I discovered this early on in the process of batching up the events, when I devised a solution which branched the dag (which is not the solution that landed). The `prev_event` for each event created in this path is passed in to `create_event` and that gets passed on down to the builder.

**Comment by H-Shay** [comment]
Right I've split up this PR, the first half deals with the auth events and is here: #14214. Once that's approved I will put up a PR dealing with the push rule changes and adding the initial power level event.

**Comment by H-Shay** [comment]
The second half is at  #14228

**Pull request closed without merging**